Spec: Add overview, add additional details to message encryption section (#227)
freedomofpress/securedrop-protocol: docs/protocol.md

## Table of contents

- [Overview](#overview)
  - [Introduction](#introduction)
  - [Sequence Diagram](#sequence-diagram)

> - [Keys]

Reviewer: Here and next, just fixing some incomplete Markdown links:

```suggestion
- [Keys](#keys)
```

Author: thank you, markdown links/headers addressed in [9397036](https://github.com/freedomofpress/securedrop-protocol/pull/227/commits/939703639cd84b3c41c5b50b4be01f9a6ef58b38)

  - [Key Hierarchy](#key-hierarchy)
  - [Key Setup Steps](#key-setup-steps)

> - [Messaging Protocol]

Reviewer: ```suggestion
- [Messaging Protocol](#messaging-protocol)
```

### Introduction

> SecureDrop Protocol is a first-contact messaging protocol between anonymous users (sources) and well-known user(s) (journalists) with a shared affiliation (newsroom).

Reviewer: Nit: I always think of "well-known" here in the sense of `/.well-known`, as in published and intentionally enumerable.  But I also trip over it every time I read it and have to convince myself of this sense all over again.  :-)

I like your "non-anonymous" below—would it be sufficient here too?

The design is largely motivated by the requirement that sources avoid local persistent state, for plausible deniability.

This specification describes:

- Each party (source, journalist, newsroom, FPF) and their [setup](#key-setup-steps)
- Message encryption, retrieval, and decryption
- What type of security and confidentiality properties are provided
- What encryption algorithms and parameters are used

Throughout, the terms **source**, **journalist** and **newsroom** are used.
These may be understood to mean: the anonymous users who initiate conversations (sources); the non-anonymous users who receive and reply to messages (journalists), and the public organization that manages the system and authorizes allowed journalists (newsroom).

The terms **sender** and **recipient** are also used, more abstractly, to refer to a user's role at a given point in the protocol's execution: when a source writes to a journalist, they are a sender, and when they receive a reply, they are a recipient, and vice-versa.

The protocol has:

> - A limited, stateless, _unauthenticated_ public API (`requestKeys`, `sendMessage`, `requestMessages`, `getMessage`) used by all parties for message sending and retrieval

Author: `requestMessages` is the term used elsewhere, I would be happy to see it changed to something clearer (requestChallenges, requestHints) but I stuck with what was already there.

Reviewer: I agree.  The *goal* is to request messages, the mechanism is to request challenges or hints; it's hard to find a good concept in the middle.  Sometimes I shorthand it as "check for messages".

- A limited, authenticated administrator API, used by newsrooms to enroll and unenroll journalists

> - A limited, authenticated journalist API, used to replenish their message encryption keys

Reviewer: It's very helpful to distinguish these.  Is it worth calling out that the authentication of these APIs/endpoints is not yet specified or implemented?

One of the system's goals is to consider real-world deployment scenarios and their risks.

> The choice of an unauthenticated API avoids a serverside "users" database.

Reviewer: ```suggestion
The choice of an unauthenticated API avoids a server-side database of user accounts.
```

#### Design Constraints

- Prioritize the safety/anonymity of the source
- Do not require sources to use any specific software or download any applications to communicate; they should be able to use Tor Browser, visit a url like `newsorg.securedrop.tor.onion`, and begin messaging
- Be maintainable: Use well-known encryption primitives and existing cryptography libraries
- Be readily-deployable: Use a single-server design, consider realistic threat models with respect to cloud deployments.

> For further context, see [our Research page](https://securedrop.org/research/), particularly the blog post series.

Reviewer: Worth including something like this there, anticipating #211?

```suggestion
For further context, see [our Research page](https://securedrop.org/research/), particularly the blog post series.  An eprint is forthcoming.
```

Concretely:

> <!-- FIXME: This is hard to read and `info` is overused and confusing. -->

Author: this is a placeholder; I think we can improve on the `info = info + ....` description and also improve on the way the SD-APKE.AuthEnc method signature kind of hides the building of the `info` param. It [bit us in the Tamarin ](https://github.com/freedomofpress/securedrop-protocol-models/blob/18bf3890b79d1f36b1ffb1b800e2d29cec5c1efe/securedrop.spthy#L244-L245) (this is a result of going by the AuthEnc method signature rather than the actual implementation).

Reviewer: Yes, I think it's time!  I've filed #233 to track this separately, unless you're eager to tack it on here.

        pk=(pkR1, pkR2),
        m, ad, info):
    (c2, K2) = KEM_PQ.Encap(pkR=pkR2)

>     (c1, cp) = pskAEnc(skS=skS1, pkR=pkR1, psk=K2, m=m, ad=ad, info=c2 + info + pkS2)  # where cp = c', `info` is receiver fetch pubkey, pkS2 is sender KEM_PQ pubkey/encaps key, from SD-APKE tuple

Reviewer 2: don't we not have `pkS2` in the scope here? we should just add a line deriving it or take it as a paramerer

Author: Ah, good point. Addressed in 19eb465 in not the most graceful way (just `skS2.public()`) but hopefully comprehensible

>     return ((c1, cp), c2)

Author: (see above -  I know this doesn't really belong here in a comment, but I think somewhere we need a clear place to explain what the authenc `info` is)

Reviewer: Yes, I bet this will need less explanation with a better name from #233.

For each recipient, a sender produces a message ciphertext (SD-APKE ciphertext), a metadata ciphertext (SD-PKE ciphertext), and a message delivery hint.

A sender knows their own keys, the newsroom's verification key $vk_{NR}^{sig}$, and
the $pks$ and $sigs$ they previously [fetched].

In addition, in the **reply case,** if the sender is a journalist replying to a
source, they also already know their recipient's keys without further
verification.
In this case, they substitute the source's long-term keys
for their own in the recipient list, and address the remaining slots to all other
enrolled journalists.

**Message Ciphertext (SD-APKE Ciphertext).** The SD-APKE ciphertext is sender authenticated using classical, DH-AKEM implicit authentication, and provides hybrid (classical/quantum-resistent) message encryption via a quantum-resistent shared secret, `pskAPKE`.
The SD-APKE ciphertext carries a [structured plaintext message](#message-formats) including the sender's long-term $fetch$ and $PKE$ public keys, which must be enclosed by the source so that they can receive replies, and are enclosed by the journalist for parity.

> Despite the name, `pskAPKE` is not a true 'pre-shared' key, and functions more like a [KEM combiner](https://datatracker.ietf.org/doc/draft-ounsworth-cfrg-kem-combiners/); the naming convention from [related work](https://eprint.iacr.org/2023/1480) is retained.

Reviewer: Just some suggestions for clarity (and to reuse citations):

```suggestion
**Message Ciphertext (SD-APKE Ciphertext).** The SD-APKE ciphertext is sender-authenticated using classical, DH-AKEM implicit authentication, and provides hybrid (post-quantum/traditional) message encryption via a quantum-resistant shared secret produced by `pskAPKE`.
The SD-APKE ciphertext carries a [structured plaintext message](#message-formats) including the sender's long-term $fetch$ and $PKE$ public keys, which must be enclosed by the source so that they can receive replies, and are enclosed by the journalist for parity.
Despite the name, `pskAPKE` does not produce a true 'pre-shared' key; it functions here more like a [KEM combiner](https://datatracker.ietf.org/doc/draft-ounsworth-cfrg-kem-combiners/).  We follow the naming convention of Alwen et al. (2023), ["The Pre-Shared Key Modes of HPKE"][alwen2023], which introduced this usage.
```

The SD-APKE ciphertext (message ciphertext) additionally commits to the following values via the HPKE `info` parameter: the encapsulation of the `pskAPKE` shared secret; the receiver's $fetch$ key, and the ML-KEM portion of the sender's SD-APKE public key.

This `info` parameter is greater than 64 bytes. Implementors MUST ensure that the HPKE implementation and the underlying AEAD support a sufficiently long `info` parameter, or implement a modification to the protocol that hashes the concatenated values to the supported `info` length.

This `info` parameter MUST NOT be transmitted with the ciphertext by the underlying AEAD, since it contains cleartext public keys, which are identifying; comformant implementations of HPKE pass the `info` parameter to the KeySchedule but do not transmit it with the ciphertext. The recipient MUST locally reconstruct the `info` parameter based on the information in the PKE ciphertext and their knowledge of their own $fetch$ key, and supply the reconstructed `info` parameter to successfully decrypt the message.

> _One-time drop mode extension_: An extension implementation MAY omit the sender's $fetch$ and $PKE$ keys from the plaintext message, offering improved deniability (no possibility for pending ciphertexts) but forgoing the sender's ability to receive replies.

Reviewer: Nice to have this here!

**Metadata Ciphertext (SD-PKE Ciphertext).** Because decrypting the SD-APKE ciphertext requires the recipient to know the sender's long-term APKE public key, an SD-PKE ciphertext (metadata ciphertext) delivers this SD-APKE public key, encrypted to the recipient's $PKE$ key, thus keeping the sender's identity hidden from the server, as described in HPKE's metadata protection guidance ([RFC 9180 §9.9]).

The SD-PKE (metadata) ciphertext is unauthenticated, so its contents MUST be committed to in the SD-APKE ciphertext's encryption context.
This is satisfied by the use of implicit authenticated encryption plus the `info` parameter, as described above.

> The SD-PKE ciphertext MUST provide hybrid classical/quantum guarantees.

Reviewer: As above, since these seem to be the canonical terms since RFC 9794 ["Terminology for Post-Quantum Traditional Hybrid Schemes"](https://datatracker.ietf.org/doc/html/rfc9794#section-5-3.2.1):

```suggestion
The SD-PKE ciphertext MUST provide hybrid post-quantum/traditional confidentiality.
```

**Message Delivery Hint.** The sender also computes a hint from the recipient's fetching key: a fresh
ephemeral DH public key $X = g^x$ and a Diffie–Hellman share $Z =
(pk_R^{fetch})^x$. This lets the recipient privately scan for their messages in
step 7 without disclosing their identity to the server. The server stores the two
ciphertexts and hint under a randomly generated message ID.

> As follows, the final message paylod to the server includes: each ciphertext; the encapsulated shared secrets required to decrypt each of them; the encapsulated shared secret of the `pskAPKE`; and the two components of the message delivery hint.

Reviewer: ```suggestion
As follows, the final message payload to the server includes: each ciphertext; the encapsulated shared secrets required to decrypt each of them; the encapsulated shared secret of the `pskAPKE`; and the two components of the message delivery hint.
```

Author: thank you for catching, addressed in [47d5322](https://github.com/freedomofpress/securedrop-protocol/pull/227/commits/47d53228f1ff3c40fff32504552017a80f0284ab) , & all other fixups

|                                                                                                  |                                                | $`fetched \gets fetched \cup \{cid\}`$                                                          |
|                                                                                                  |                                                | If $`tofetch \setminus \{cid\} \neq \emptyset`$: repeat from `RequestMessages`                  |

Implementors MUST mitigate timing attacks via the API that could leak the number of ciphertexts on the server, for example by ensuring that `requestMessages` is constant-time at the server.

> 

Author: (To discuss: OK?)

### Message Formats

Implementors MUST implement robust message-parsing and are expected to gracefully handle malformed plaintext and ciphertext messages, both at the server and on the client.

### Plaintext

#### SD-APKE (Message) Plaintext

> <!-- FIXME: protocol versioning?; message padding -->

Author: (We haven't talked about whether we will specify a version string for the protocol itself, and if so, where it will live.)

Reviewer: Could cross-reference #228 explicitly here?

`SENDER_FETCH_PUBKEY_BYTES || SENDER_PKE_PUBKEY_BYTES || structured_message_bytes || padding`

Unpadded len: 32 bytes + 1216 bytes + len(structured_message)

= structured message size + 1248; messages will then be padded to a fixed message size before encryption.

#### SD-PKE (Metadata) Plaintext

<!-- FIXME: protocol versioning? -->

`SENDER_LONG_TERM_SD_APKE_DHAKEM_PUBKEY_BYTES || SENDER_LONG_TERM_SD_APKE_MLKEM768_PUBKEY_BYTES`

Len: 32 + 1184 = 1216 bytes

### Ciphertext

Encrypting the SD-APKE (message) plaintext yields a ciphertext and encapsulated shared secret ciphertext:

`CT_APKE` = `MLKEM768_CT_ENCAPS_SHARED_SECRET || DHAKEM_ENCAPS_SHARED_SECRET || CT_SD_APKE`

Len: 1088 + 32 + ((fixed message size) + AEAD_TAG_LEN) = 1120 + (fixed message size + 16)

= fixed message size + 1136

Encrypting the SD-PKE (metadata) plaintext also yields a ciphertext and encapsulated shared secret ciphertext:

`CT_PKE` = `CT_SD_PKE_ENCAPS_SS_CT || CT_SD_PKE`

> Len: XWING_SHARED_SECRET_ENCAPS_CT_LEN + (DHAKEM_PK_LEN + MLKEM768_PK_LEN + AEAD_TAG_LEN) = 1120 + 1184 + 16

Reviewer 2: this adds up to 2320, just missing the `DHAKEM_PK_LEN` to get to 2352:

```suggestion
Len: XWING_SHARED_SECRET_ENCAPS_CT_LEN + (DHAKEM_PK_LEN + MLKEM768_PK_LEN + AEAD_TAG_LEN) = 1120 + 32 + 1184 + 16
```

Author: ugh, thank you - addressed in 19eb465 (and all such)

= 2352

### Encrypted Envelope (Message payload)

`encrypted_envelope` = `X || Z || CT_APKE || CT_PKE` = (epehemeral pk || dh(ephemeral pk, receiver fetch pk) || CT_APKE || CT_PKE)

Len: 32 + 32 + (fixed message size + 1136) + 2352

= fixed message size + 3552

### Message storage on server

Message tuples:
`(server_generated_uuid, encrypted_envelope, challenge_pk, dh_share, fuzzy_exp_timestamp)`

> Len: 16, len(`encrypted_envelope`), 32, 32, 12 * n total server capacity (server pads )

Reviewer 2: i'm having trouble parsing `12 * n total server capacity (server pads )`?

Author: it was terrible, fixed it up :)

### Message delivery hint (Challenges)

> `(AEADenc(server_generated_uuid), server_dh_share, mgdh)`

Reviewer 2: where possible if we have a variable name defined can we use the same nomenclature? just to make this easier to track, e.g. i know what mgdh maps to from the previous code, but it's not obvious

Len: 32 + 32 + 32 = 96 bytes * n challenges

## Known Limitations

- The protocol does not currently include a specification for transferring attachments.
- The protocol does not currently include a specification for Journalist key replenishment, or for rotation of journalist long-term keys.
- The protocol does not currently include a specification for rotation of the Newsroom key. The relationship between the Newsroom key and the server URL is not yet specified.
- The protocol is not designed for scalability. There is a maximum number of messages that can be held by the server, constrained by the number of per-request challenges that the server can reasonably perform during message-fetching without unacceptable latency for users, particularly over Tor. See benchmarks for more information.
- The use of HPKE's implicit authentication for message sending means that the protocol is vulnerable to [key compromise impersonation](https://datatracker.ietf.org/doc/html/rfc9180#section-9.1.1).

> - The protocol currently offers quantum-resistent message encryption, but not quantum-resistent message authentication or message-fetching.

Reviewer: ```suggestion
- The protocol currently offers quantum-resistant message encryption, but not quantum-resistant message authentication or message-fetching.
```
